### PR TITLE
group varying factor levels fixed

### LINKED ISSUE #2199: malformed factor resulting from 'by' expression when using melt.data.table in 'j' expression
I get unprintable data.table object whose printing results in error 
```
Error in as.character.factor(x) : malformed factor
```
but in some cases (large dataset) also crashes R.

My code looks like this
```
require(data.table)

# generate demo dataset
ids   <- letters[1:3]
dates <- 1:2

dt <- data.table(CJ(dates, ids, ids))
setnames(dt, c("date", "id1", "id2"))

dt[, value := rnorm(18)]

# IMPORTANT: to reproduce the bug, drop some id
dt <- dt[!(date == 1 & (id1 == "a" | id2 == "a"))]

# demo function
f1 <- function(sdt) {
  dt1 <- dcast.data.table(sdt, id1 ~ id2)
  dt2 <- melt.data.table(dt1, id.vars = "id1")
  print(dt2)
  
  dt2
}

res <- dt[, f1(.SD), by = date]
#id1 variable      value
#1:   b        b -1.3643635
#2:   c        b  0.5305674
#3:   b        c  0.2023935
#4:   c        c  0.1063894
#id1 variable       value
#1:   a        a -0.35193161
#2:   b        a -0.65570503
#3:   c        a  0.01524152
#4:   a        b -0.07234880
#5:   b        b -1.16267653
#6:   c        b -1.41490080
#7:   a        c  0.10225144
#8:   b        c -0.84277336
#9:   c        c -0.23164772

res
#Error in as.character.factor(x) : malformed factor
```

In my case I forgot to put the usual `variable.factor = FALSE` into the melt.data.table() **which makes it work ok**. But this behavior surprises me. It only appears when the two sets of factors differ (for date=1 and date=2 the ids are different sets), so if you I skip the line
```
dt <- dt[!(date == 1 & (id1 == "a" | id2 == "a"))]
```
it works alright.

I am on the latest stable version of data.table.

My sessionInfo()
```
R version 3.4.0 (2017-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: KDE neon User Edition 5.10

...

other attached packages:
[1] data.table_1.10.4

loaded via a namespace (and not attached):
[1] compiler_3.4.0 tools_3.4.0
```

### LINKED ISSUE #2522: Unexpected result when creating a factor column with 'by=' present
For example,

```
library(data.table)

dt <- data.table(
  id = 1:9,
  grp = rep(1:3, each = 3),
  val = c("a","b","c",   "a","b","c",   "a","b","c")
)

dt[, valfactor1 := factor(val), by = grp]  # fine and dandy
dt[, valfactor2 := factor(val), by = id]   # breaks
```

```
dt
   id grp val valfactor1 valfactor2
1:  1   1   a          a          c
2:  2   1   b          b          c
3:  3   1   c          c          c
4:  4   2   a          a          c
5:  5   2   b          b          c
6:  6   2   c          c          c
7:  7   3   a          a          c
8:  8   3   b          b          c
9:  9   3   c          c          c
```

I suspect this has come up before, but I didn't see anything on it.  Also, I'm not sure what a reasonable behavior would be if I was trying to assign factors by group with ordered levels.  At the least, I think a warning in these cases would be helpful.  Right now it silently does something strange.

Thanks!

```
sessionInfo()

R version 3.4.2 (2017-09-28)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.10.5

loaded via a namespace (and not attached):
[1] compiler_3.4.2 tools_3.4.2    yaml_2.1.15
```

### PR DESCRIPTION
**Pull request by mattdowle**
Closes #2199 
Closes #2522

### CODE REVIEW

File: src/assign.c [hunk 1]
```diff
@@ -818,6 +818,88 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
       source = PROTECT(copyAsPlain(source)); protecti++;
     }
   }
+  const bool sourceIsFactor=isFactor(source), targetIsFactor=isFactor(target);
+  if (sourceIsFactor && targetIsFactor) {
+    // TODO:         ^^ could be || in future; assign is already handling that so leave that for now. Then move more of assign to call here.
+    if (!sourceIsFactor || !targetIsFactor) {
```
[the remark this answers is not in the document]
Comment (mattdowle) [reply src/assign.c]: it's there for the future when the todo in that comment is done

[the remark this answers is not in the document]
Comment (jangorecki) [reply src/assign.c]: make sense, but could be commented out code block too.
would be good to have a followup issue and link it from there

[the remark this answers is not in the document]
Comment (mattdowle) [reply src/assign.c]: It could be but I'd prefer not to. Ok?

[the remark this answers is not in the document]
Comment (mattdowle) [reply src/assign.c]: Of all the lines in this PR, this is the least significant to pick up on.

**Pull request merged**